### PR TITLE
New Adapter: Connatix

### PR DESCRIPTION
**Pull request by patrickszeleczki**
Docs: https://github.com/prebid/prebid.github.io/pull/5595

### CODE REVIEW

File: static/bidder-info/connatix.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+endpoint: "https://capi.connatix.com/rtb/ortb"
+maintainer:
+  email: "pubsolutions@connatix.com"
```
Comment (przemkaczmarek) [inline static/bidder-info/connatix.yaml:3]: send email for verification, please respond

**Comment by przemkaczmarek** [comment]
@patrickszeleczki Please add more JSON tests (Check which lines of code are not covered by tests). We usually approve PRs when the test coverage is more than 85%.

File: static/bidder-info/connatix.yaml (hunk 1 above)
Comment (patrickszeleczki) [reply static/bidder-info/connatix.yaml:3]: done

File: static/bidder-info/connatix.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    mediaTypes:
+      - banner
+      - video
+userSync:
+  iframe:
+    url: "https://capi.connatix.com/us/pixel?pId=53&gdpr={{.GDPR}}&gdpr_consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&callback={{.RedirectURL}}"
+    userMacro: "[UID]"
+  redirect:
+    url: "https://capi.connatix.com/us/pixel?pId=52&gdpr={{.GDPR}}&gdpr_consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&callback={{.RedirectURL}}"
+    userMacro: "[UID]"
```
Comment (przemkaczmarek) [inline static/bidder-info/connatix.yaml:20]: only iframe works:
image

Comment (patrickszeleczki) [reply static/bidder-info/connatix.yaml:20]: thanks for the feedback! should be fixed now

**Review by przemkaczmarek — APPROVED** [review]

**Review by karimMourra — APPROVED** [review]

**Comment by patrickszeleczki** [comment]
ping @bsardo 
thank you!

**Review by przemkaczmarek — DISMISSED** [review]

File: adapters/connatix/connatix.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+	}
+	return resArr, errs
+}
+
+func buildRequestImp(imp *openrtb2.Imp, ext impExtIncoming, displayManagerVer string, reqInfo *adapters.ExtraRequestInfo) error {
+	if imp.Video == nil && imp.Banner == nil {
+		return &errortypes.BadInput{
+			Message: "Either video or banner object on impression is required",
+		}
+	}
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go]: handled in [8d50d87](https://github.com/Connatix/prebid-server/commit/8d50d8780c67a352296a55de210c23fc4f11b87e)

File: adapters/connatix/connatix.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if request.Site == nil && request.App == nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Either site or app object is required",
+		}}
+	}
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go]: handled in [8d50d87](https://github.com/Connatix/prebid-server/commit/8d50d8780c67a352296a55de210c23fc4f11b87e)

File: adapters/connatix/connatix.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+	// connatix adapter expects imp.displaymanagerver to be populated in openrtb2 request
+	// but some SDKs will put it in imp.ext.prebid instead
+	displayManagerVer := buildDisplayManageVer(request)
+
+	var errs []error
+
+	validImps := []openrtb2.Imp{}
+	for i := 0; i < len(request.Imp); i++ {
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go]: handled in [967e769](https://github.com/Connatix/prebid-server/commit/967e76967c9f04b35e46a32eedc73d84bd562f3e)

File: adapters/connatix/connatix.go [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+		}
+
+		validImps = append(validImps, request.Imp[i])
+	}
+	request.Imp = validImps
+
+	// If all the requests were malformed, don't bother making a server call with no impressions.
+	if len(request.Imp) == 0 {
+		return nil, errs
+	}
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go]: handled in [967e769](https://github.com/Connatix/prebid-server/commit/967e76967c9f04b35e46a32eedc73d84bd562f3e)

File: adapters/connatix/connatix.go [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+		}
+
+		if err := buildRequestImp(&request.Imp[i], impExtIncoming, displayManagerVer, reqInfo); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		validImps = append(validImps, request.Imp[i])
+	}
+	request.Imp = validImps
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go]: handled in [967e769](https://github.com/Connatix/prebid-server/commit/967e76967c9f04b35e46a32eedc73d84bd562f3e)

File: adapters/connatix/connatix.go [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+	source, err := jsonparser.GetString(req.App.Ext, openrtb_ext.PrebidExtKey, "source")
+	if err != nil {
+		return ""
+	}
+
+	version, err := jsonparser.GetString(req.App.Ext, openrtb_ext.PrebidExtKey, "version")
+	if err != nil {
+		return ""
+	}
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go:231]: handled in [181c4f9](https://github.com/Connatix/prebid-server/commit/181c4f94ce4a1ae0b0faa99da3c956189262eef8)

File: adapters/connatix/connatix.go [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+
+	return fmt.Sprintf("%s-%s", source, version)
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go:233]: handled in [181c4f9](https://github.com/Connatix/prebid-server/commit/181c4f94ce4a1ae0b0faa99da3c956189262eef8)

File: adapters/connatix/connatix.go [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+	}
+
+	// Divide imps to several requests
+	requests, errors := splitRequests(request.Imp, request, a.uri.String())
+	return requests, append(errs, errors...)
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(response) {
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go:73]: handled in [181c4f9](https://github.com/Connatix/prebid-server/commit/181c4f94ce4a1ae0b0faa99da3c956189262eef8)

File: adapters/connatix/connatix.go [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+	}
+
+	if err := adapters.CheckResponseStatusCodeForErrors(response); err != nil {
+		return nil, []error{err}
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go:77]: handled in [181c4f9](https://github.com/Connatix/prebid-server/commit/181c4f94ce4a1ae0b0faa99da3c956189262eef8)

File: adapters/connatix/connatix.go [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+	}
+
+	var connatixResponse openrtb2.BidResponse
+	if err := jsonutil.Unmarshal(response.Body, &connatixResponse); err != nil {
+		return nil, []error{err}
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go:82]: handled in [181c4f9](https://github.com/Connatix/prebid-server/commit/181c4f94ce4a1ae0b0faa99da3c956189262eef8)

File: adapters/connatix/connatix.go [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+	if len(imp.DisplayManagerVer) == 0 && len(displayManagerVer) > 0 {
+		imp.DisplayManagerVer = displayManagerVer
+	}
+
+	// Check if imp comes with bid floor amount defined in a foreign currency
+	if imp.BidFloor > 0 && imp.BidFloorCur != "" && !strings.EqualFold(imp.BidFloorCur, "USD") {
+		// Convert to US dollars
+		convertedValue, err := reqInfo.ConvertCurrency(imp.BidFloor, imp.BidFloorCur, "USD")
+		if err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go:198]: handled in [181c4f9](https://github.com/Connatix/prebid-server/commit/181c4f94ce4a1ae0b0faa99da3c956189262eef8)

File: adapters/connatix/connatix.go [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+
+	bidderResponse.Currency = "USD"
+
+	return bidderResponse, errs
+}
+
+func validateAndBuildImpExt(imp *openrtb2.Imp) (impExtIncoming, error) {
+	var ext impExtIncoming
+	if err := jsonutil.Unmarshal(imp.Ext, &ext); err != nil {
+		return impExtIncoming{}, err
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go:114]: handled in [181c4f9](https://github.com/Connatix/prebid-server/commit/181c4f94ce4a1ae0b0faa99da3c956189262eef8)

File: adapters/connatix/connatix.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go]: handled in [181c4f9](https://github.com/Connatix/prebid-server/commit/181c4f94ce4a1ae0b0faa99da3c956189262eef8)

**Comment by patrickszeleczki** [comment]
@bsardo Thanks a lot for the thorough review and I really appreciate the clear suggestions for the new tests and the small improvement suggestions! 🙇

**Comment by patrickszeleczki** [comment]
ping @bsardo @przemkaczmarek 
thank you!

**Review by bsardo — COMMENTED** [review]
Looks good, just a couple more comments.

File: adapters/connatix/connatix.go [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v3/adapters"
+	"github.com/prebid/prebid-server/v3/config"
+	"github.com/prebid/prebid-server/v3/errortypes"
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+	"github.com/prebid/prebid-server/v3/util/jsonutil"
+)
+
+const (
+	maxImpsPerReq = 1
```
Comment (przemkaczmarek) [inline adapters/connatix/connatix.go:20]: Why do you have a value of 1 here?

File: adapters/connatix/connatix.go [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+		uri: *uri,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if request.Device != nil && request.Device.IP == "" && request.Device.IPv6 == "" {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Device IP is required",
+		}}
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go]: indeed. thank you! fixed

File: adapters/connatix/models.go [hunk 17]
```diff
@@ -0,0 +1,31 @@
+package connatix
+
+import (
+	"net/url"
+
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+)
+
+type adapter struct {
+	uri url.URL
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/models.go]: fixed

File: adapters/connatix/connatix.go [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	uri, err := url.Parse(config.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	bidder := &adapter{
+		uri: *uri,
+	}
+	return bidder, nil
```
[the remark this answers is not in the document]
Comment (patrickszeleczki) [reply adapters/connatix/connatix.go]: fixed

**Review by przemkaczmarek — APPROVED** [review]

**Review by bsardo — APPROVED** [review]

**Comment by patrickszeleczki** [comment]
@bsardo @przemkaczmarek do you have any estimate when this will be included in a release? thank you!

**Pull request merged**